File: Data Structures & Algorithms/word-break/submission-2.cpp
class Solution {
public:
    // bool solve(string s, int index, vector<bool>& dp, vector<string>& wordDict){
    //     if(index >= s.size()){
    //         return true;
    //     }
    //     if(dp[index]) return dp[index];
    //     for(auto word: wordDict){
    //         if(s.substr(index,word.size()) == word){
    //             dp[index] = solve(s, index+word.size(), dp, wordDict);
    //         }
    //     }
    //     return dp[index];
    // }
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<bool> dp(s.size()+1, false);
        dp[s.size()] = true;
        for(int i = s.size()-1; i>=0; i--){
            for(auto word: wordDict){
                if(i + word.size() <= s.size() && s.substr(i, word.size()) == word){
                    dp[i] = dp[i+ word.size()];
                }
                if(dp[i]){
                    break;
                }
            }
        }
        return dp[0];
    }
};

File: Data Structures & Algorithms/search-for-word/submission-2.py
class Solution:
    def solve(self, row:int, col:int, index:int, word:str, board:list[list[str]]) -> bool:
        if index >= len(word):
            return True
        if row < 0 or col < 0 or row >= len(board) or col >= len(board[0]) or board[row][col] != word[index]:
            return False
        c = board[row][col]
        board[row][col] = '!'
        delr = [-1, 0, 1, 0]
        delc = [0, -1, 0, 1]
        for i in range(0,4):
            nrow = row + delr[i]
            ncol = col + delc[i]
            if self.solve(nrow, ncol, index+1, word, board):
                return True
        board[row][col] = c
        return False
    def exist(self, board: list[list[str]], word: str) -> bool:
        n,m = len(board), len(board[0])
        for i in range(0,n):
            for j in range(0, m):
                if self.solve(i,j,0,word,board):
                    return True
        return False
    
